### PR TITLE
Build script improvements

### PR DESCRIPTION
**Pull request by themissingcow**
See individual commits, but headline news:

 - `build.py` now uses an existing docker image build (either local or remote)
 - `build-docker.py` can be used to build/publish docker images
 - `ARNOLD_ROOT` and `DELIGHT_ROOT` are now optional, build of those component will be skipped.
 -  The Centos version, along with`yum` and `pip` package versions are locked.
 - `build.py` is no longer included in the container to reduce the maintenance overhead of publish images.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,11 +2,31 @@
...
+COPY yum-versionlock.list /etc/yum/pluginconf.d/versionlock.list
+
+RUN yum install -y yum-versionlock && \
+	./versionlock.sh list-installed /tmp/packages && \
+#
+#
+# NOTE: If you add a new yum package here, make sure you update the version
+# lock files as follows and commit the changes to yum-versionlock.list:
+#
+#   ./build.py --project build --update-package-versions 1 --new-only 1
```
Comment (johnhaddon) [inline Dockerfile]: Needs update

File: README.md [hunk 2]
```diff
@@ -43,3 +43,60 @@ Remove stopped containers :
...
+> NOTE: The `build-docker.py` uses the simpler flag syntax for most of it options,
+> this means its a little different to `build.py`, we aim to update that over
+> time to match.
+
+In order to produce this image, we make heavy use of `yum` to install the
+packages required to build Gaffer and it's dependencies. As we don't manually
+specify every version of every package we risk a non-deterministic build
+environment. To get around this, we make use of `yum versionlock`. The
+`yum-versionlock.list` file in the repository is copied into the base Centos
+image such that when `yum` runs, it will repeatable install the expected
```
Comment (johnhaddon) [inline README.md]: repeatable -> repeatably

File: build.py [hunk 3]
```diff
@@ -120,22 +134,24 @@
 
 # Check that our environment contains everything we need to do a build.
 
-for envVar in ( "GITHUB_RELEASE_TOKEN", ) :
-	if envVar not in os.environ	:
-		parser.exit( 1,  "{0} environment variable not set".format( envVar ) )
+if args.upload :
+	if "GITHUB_RELEASE_TOKEN" not in os.environ	:
+		parser.exit( 1,  "GITUHB_RELEASE_TOKEN environment variable not set" )
```
Comment (johnhaddon) [inline build.py]: Add a check that Arnold and 3Delight root are set, and exit if not. That way we can't accidentally make a release build without full renderer support.

**Review by johnhaddon — CHANGES_REQUESTED** [review]
Nice! Could you fix up those couple of little things and then merge?

File: build.py (hunk 3 above)
Comment (themissingcow) [reply build.py]: Added in https://github.com/GafferHQ/build/pull/10/commits/683e03b84fcbb9f69ba56155d09f42aaa0903962

**Pull request merged**